src/server.rs

```diff
@@ -149,23 +149,28 @@ impl UdpServer {
     async fn gossip_multiple(&self, rng: &mut SmallRng) {
         let scuttlebutt = self.scuttlebutt.lock().await;
         let self_node_id = &scuttlebutt.self_node_id;
-        let mut rand_nodes = [""; GOSSIP_COUNT];
+        const EMPTY_STRING: String = String::new();
+        let mut rand_nodes = [EMPTY_STRING; GOSSIP_COUNT];
 
         // Select up to [`GOSSIP_COUNT`] node IDs at random.
         let nodes = scuttlebutt.cluster_state_map.nodes();
         let count = nodes
             .filter(|node_id| node_id != &self_node_id)
-            .map(|node_id| node_id.as_str())
+            .cloned()
             .choose_multiple_fill(rng, &mut rand_nodes);
 
+        // Drop lock to prevent deadlock in [`UdpSocket::gossip`].
+        drop(scuttlebutt);
```

Comment:
Now this is kinda interesting. After you mentioned potential locking issues I immediately added a comment to ensure that the lock created in `gossip_multiple` doesn't create issues with the lock in `gossip`.

Reading the tokio documentation it sounded like holding a lock across await boundaries wasn't that big of a deal, since that seems to be its main benefit over STD's lock. Unfortunately the behavior isn't really documented precisely, but after writing the `heartbeat` test I found out that this **will** cause a deadlock.

So unfortunately this allocates strings now, but at least it doesn't deadlock which is *probably* more important.